stored_file.php

```diff
@@ -0,0 +1,387 @@
+
+    public $id;
+    public $extension;
+    public $mime;
+    public $filename;
+    public $source;
+
+    protected $__moved;
+
+    public function actual_filename()
+    {
+        if ($this->__moved)
+            return static::resolve_file($this->filename);
+        return $this->source;
+    }
+
+    public function has_supported_extension($supported_exts = null)
+    {
+        if (!$this->extension)
+            $this->extension = static::parse_extension($this->filename);
+
+        if (!$this->extension)
+            return false;
+
+        if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+        {
+            return in_array($this->extension,
+                $supported_exts);
+        }
+
+        return in_array($this->extension,
+            static::$__supported_exts);
+    }
+
+    public function exists()
+    {
+        return is_file($this->actual_filename());
+    }
+
+    public function size()
+    {
+        return filesize($this->actual_filename());
+    }
+
+    public function human_size()
+    {
+        $size = $this->size();
+
+        if ($size > static::SIZE_GIB)
+        {
+            $size = ($size / static::SIZE_GIB);
+            return sprintf('%.2f GiB', $size);
+        }
+
+        if ($size > static::SIZE_MIB)
+        {
+            $size = ($size / static::SIZE_MIB);
+            return sprintf('%.2f MiB', $size);
+        }
+
+        $size = ($size / static::SIZE_KIB);
+        return sprintf('%.2f KiB', $size);
+    }
+
+    public function detect_mime()
+    {
+        if ($this->mime) return $this->mime;
+        $this->mime = File_Util::detect_mime($this->actual_filename());
+        return $this->mime;
+    }
+
+    public function generate_filename($ext = null)
+    {
+        if (!$ext) $ext = $this->extension;
+
+        $hash = md5_file($this->source);
+        $dir1 = substr($hash, 0, 2);
+        $dir2 = substr($hash, 2, 2);
+        $name = substr($hash, 4, 28);
+
+        $this->filename = build_path($dir1, $dir2,
+            sf('%s.%s', $name, $ext));
+        return $this->filename;
+    }
+
+    public function read()
+    {
+        return file_get_contents($this->actual_filename());
+    }
+
+    public function move($invoker_tag = null)
+    {
+        if ($this->__moved) return;
+
+        if (!$invoker_tag) $invoker_tag = md5(serialize(debug_backtrace(5)));
+
+        $destination = static::resolve_file($this->filename, true);
+        $dir = dirname($destination);
+        $ref_filename = $destination . '.ref';
+
+        if (!is_dir($dir)) mkdir($dir, 0755, true);
+
+        if (!file_exists($destination)) {
+
+            rename($this->source, $destination);
+
+            // Create a new file reference
+            $ref_res = fopen($ref_filename, 'w');
+            if (flock($ref_res, LOCK_EX)) {
+                ftruncate($ref_res, 0);
+                fwrite($ref_res, "{\"ref_count\": 1,\"atags\" : [\"$invoker_tag\"],\"dtags\" : []}");
```

Comment:
This is an issue. You are writing JSON in an unsafe way. It's like an SQL injection... but with JSON. If another developer was to pass in a string that contained a double quote `"` the JSON would no longer be valid. It's not clear to me why you did this manually this instead of using `json_encode`.

Reply:
Yes, you are right, we can use json_encode to escape invoker_tag, I did not use it because this variable should not be a user input, it must be a label carefully chosen by the developer according to the location where he uses the corresponding method.